xdsl/parser_ng.py

```diff
@@ -0,0 +1,908 @@
+
+        """
+        save = self.save()
+
+        if break_on is not None:
+            self.break_on = break_on
+        if ignore_whitespace is not None:
+            self.ignore_whitespace = ignore_whitespace
+
+        try:
+            yield self
+        finally:
+            self.break_on = save[1]
+            self.ignore_whitespace = save[2]
+
+
+class ParserCommons:
+    """
+    Colelction of common things used in parsing MLIR/IRDL
+
+    """
+    integer_literal = re.compile(r'[+-]?([0-9]+|0x[0-9A-f]+)')
+    decimal_literal = re.compile(r'[+-]?([1-9][0-9]*)')
+    string_literal = re.compile(r'"([^\n\f\v\r"]|\\[nfvr"])+"')
+    float_literal = re.compile(r'[-+]?[0-9]+\.[0-9]*([eE][-+]?[0-9]+)?')
+    bare_id = re.compile(r'[A-z_][A-z0-9_$.]+')
+    value_id = re.compile(r'%[A-z_][A-z0-9_$.]+')
+    suffix_id = re.compile(r'([0-9]+|([A-z_$.-][0-9A-z_$.-]*))')
+    block_id = re.compile(r'\^([0-9]+|([A-z_$.-][0-9A-z_$.-]*))')
+    type_alias = re.compile(r'![A-z_][A-z0-9_$.]+')
+    attribute_alias = re.compile(r'#[A-z_][A-z0-9_$.]+')
+    builtin_type = re.compile('({})'.format(
+        '|'.join((
+            r'[su]?i\d+', 'tensor', 'vector',
+            'memref', 'complex', 'opaque',
+            'tuple', 'index',
+            # TODO: add all the FloatNtype, Float8E4M3FNType, Float8E5M2Type, and BFloat16Type
+        ))
+    ))
+    double_colon = re.compile('::')
+    comma = re.compile(',')
+
+    class BNF:
+        """
+        Collection of BNF trees.
+        """
+        generic_operation = BNF.Group([
+            BNF.Nonterminal('string-literal', bind="name"),
+            BNF.Literal('('),
+            BNF.ListOf(BNF.Nonterminal('value-id'), bind='args'),
+            BNF.Literal(')'),
+            BNF.OptionalGroup([
+                BNF.Literal('['),
+                BNF.ListOf(BNF.Nonterminal('block-id'), allow_empty=False, bind='blocks'),
+                # TODD: allow for block args here?! (accordin to spec)
+                BNF.Literal(']')
+            ], bind='blocks_group'),
+            BNF.OptionalGroup([
+                BNF.Literal('('),
+                BNF.ListOf(BNF.Nonterminal('region'), bind='regions', allow_empty=False),
+                BNF.Literal(')')
+            ], bind='region_group'),
+            BNF.Nonterminal('attr-dict', bind='attributes'),
+            BNF.Literal(':'),
+            BNF.Nonterminal('function-type', bind='type_signature')
+        ])
+        region = BNF.Group([
+            BNF.Literal('{'),
+            BNF.ListOf(BNF.Nonterminal('operation'), separator=re.compile('')),
+            BNF.Literal('}'),
+        ])
+
+
+class MlirParser:
+    """
+    Basic recursive descent parser.
+
+    methods marked try_... will attempt to parse, and return None if they failed. If they return None
+    they must make sure to restore all state.
+
+    methods marked must_... will do greedy parsing, meaning they consume as much as they can. They will
```

Comment:
"must parse" reads like a boolean expression to me, it took a little while to understand that it was actually parsing the text. Maybe replace with either `parse` or `parse_greedy` to make into an active verb phrase?